helix-core/src/main/java/org/apache/helix/controller/stages/BestPossibleStateCalcStage.java

```diff
@@ -193,15 +193,19 @@ public Object call() {
   }
 
   // Check whether the offline/disabled instance count in the cluster reaches the set limit,
-  // if yes, pause the rebalancer, and throw exception to terminate rebalance cycle.
+  // if yes, auto enable maintenance mode, and use the maintenance rebalancer for this pipeline.
   private boolean validateOfflineInstancesLimit(final ResourceControllerDataProvider cache,
       final HelixManager manager) {
     int maxOfflineInstancesAllowed = cache.getClusterConfig().getMaxOfflineInstancesAllowed();
     if (maxOfflineInstancesAllowed >= 0) {
       int offlineCount = cache.getAllInstances().size() - cache.getEnabledLiveInstances().size();
       if (offlineCount > maxOfflineInstancesAllowed) {
+        // Enable maintenance mode in cache so the maintenance rebalancer is used for this pipeline
+        cache.enableMaintenanceMode();
```

Comment:
IMO, we should ensure the ZK update is done before enable it in the cache. This will help to prevent multiple potential consistency issues.

Reply:
I think I didn't present my idea clear enough. Since either way works for now, please put the cache update after the ZK update logic.
The reason is that we may want to change our logic that we continuously use the same cache object to optimize performance. If that is done, then the current logic is a potential bug.
Please change it if there is no other concern.